### PR TITLE
Bug in to_datetime raising ValueError with None and NaT and more than 50 elements

### LINKED ISSUE #39882: BUG: InvalidIndexError: Reindexing only valid with uniquely valued Index objects on to_datetime
#### Problem description
I am expriencing a weird bug while trying to convert a list to datetime.
Here is a bug repro:
```python
import pandas as pd
s = pd.Series([None, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.Timestamp('2016-06-07 20:07:42'), pd.NaT, pd.Timestamp('2016-05-04 20:09:22'), pd.Timestamp('2016-04-12 20:07:40'), pd.Timestamp('2016-03-30 20:10:39'), pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.Timestamp('2015-12-18 20:14:06'), pd.Timestamp('2015-12-15 20:07:59'), pd.Timestamp('2015-11-17 20:09:31'), pd.Timestamp('2015-11-05 20:10:41'), pd.Timestamp('2015-10-29 20:12:12'), pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.Timestamp('2015-07-20 20:10:23'), pd.Timestamp('2015-06-16 20:07:24'), pd.Timestamp('2015-05-28 20:09:39'), pd.NaT, pd.NaT, pd.Timestamp('2015-04-15 20:12:54'), pd.NaT, pd.NaT, pd.NaT, pd.Timestamp('2015-02-20 20:12:07'), pd.Timestamp('2014-12-29 20:04:30'), pd.Timestamp('2014-12-15 20:09:24'), pd.NaT, pd.NaT, pd.Timestamp('2014-11-17 20:03:56'), pd.NaT, pd.Timestamp('2014-10-14 20:04:47'), pd.NaT, pd.NaT, pd.NaT, pd.NaT, pd.Timestamp('2014-08-21 20:06:15'), pd.NaT, pd.Timestamp('2012-07-26 00:10:29')], dtype="object")
pd.to_datetime(s, errors="coerce")
```
What I got is :
```
InvalidIndexError: Reindexing only valid with uniquely valued Index objects
```
Testing on other example, the behavior is working as expected:

```python
import pandas as pd
s = pd.Series([None, pd.NaT, pd.Timestamp('2014-11-17 20:03:56')], dtype="object")
pd.to_datetime(s, errors="coerce")
```
gives me 
```
0                   NaT
1                   NaT
2   2014-11-17 20:03:56
dtype: datetime64[ns]
```
#### Output of ``pd.show_versions()``
<details>

INSTALLED VERSIONS
------------------
commit           : 7d32926db8f7541c356066dcadabf854487738de
python           : 3.7.4.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 19.6.0
Version          : Darwin Kernel Version 19.6.0: Tue Nov 10 00:10:30 PST 2020; root:xnu-6153.141.10~1/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : en_US.UTF-8
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.2.2
numpy            : 1.19.4
pytz             : 2020.4
dateutil         : 2.8.1
pip              : 20.2.4
setuptools       : 40.8.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : 7.19.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : 0.8.5
fastparquet      : None
gcsfs            : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 2.0.0
pyxlsb           : None
s3fs             : 0.2.2
scipy            : 1.5.4
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : 2.0.1
xlwt             : None
numba            : None

</details>

Info while scrolling a bit with pdb:
```
> /Users/jfournier/.pyenv/versions/3.7.4/lib/python3.7/site-packages/pandas/core/tools/datetimes.py(803)to_datetime()
    801         cache_array = _maybe_cache(arg, format, cache, convert_listlike)
    802         if not cache_array.empty:
--> 803             result = arg.map(cache_array)
    804         else:
    805             values = convert_listlike(arg._values, format)

ipdb> cache_array
NaT                                   NaT
NaT                                   NaT
2016-06-07 20:07:42   2016-06-07 20:07:42
2016-05-04 20:09:22   2016-05-04 20:09:22
2016-04-12 20:07:40   2016-04-12 20:07:40
2016-03-30 20:10:39   2016-03-30 20:10:39
2015-12-18 20:14:06   2015-12-18 20:14:06
2015-12-15 20:07:59   2015-12-15 20:07:59
2015-11-17 20:09:31   2015-11-17 20:09:31
2015-11-05 20:10:41   2015-11-05 20:10:41
2015-10-29 20:12:12   2015-10-29 20:12:12
2015-07-20 20:10:23   2015-07-20 20:10:23
2015-06-16 20:07:24   2015-06-16 20:07:24
2015-05-28 20:09:39   2015-05-28 20:09:39
2015-04-15 20:12:54   2015-04-15 20:12:54
2015-02-20 20:12:07   2015-02-20 20:12:07
2014-12-29 20:04:30   2014-12-29 20:04:30
2014-12-15 20:09:24   2014-12-15 20:09:24
2014-11-17 20:03:56   2014-11-17 20:03:56
2014-10-14 20:04:47   2014-10-14 20:04:47
2014-08-21 20:06:15   2014-08-21 20:06:15
2012-07-26 00:10:29   2012-07-26 00:10:29
dtype: datetime64[ns]
```
it seems that there is a deduplication which distinguishes `None` from `NaT`, then `None` got converted  to `NaT`, and ultimately the `self.is_unique` check fails as `NaT` is present twice

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #39882
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

None and NaT are different for unique while convert_listlike casts None to NaT, hence causing dups
Not sure if we could do something better.

### CODE REVIEW

File: pandas/tests/series/methods/test_convert_dtypes.py [hunk 1]
```diff
@@ -225,3 +225,14 @@ def test_convert_bool_dtype(self):
         # GH32287
         df = pd.DataFrame({"A": pd.array([True])})
         tm.assert_frame_equal(df, df.convert_dtypes())
+
+    def test_convert_object_to_datetime_with_cache(self):
+        # GH#39882
+        ser = pd.Series(
+            [None] + [pd.NaT] * 50 + [pd.Timestamp("2012-07-26")], dtype="object"
```
Comment (jbrockmendel) [inline pandas/tests/series/methods/test_convert_dtypes.py:232]: could make 50 parameter in tools.datetimes, ensures that the test stays valid even if that gets changed

Comment (phofl) [reply pandas/tests/series/methods/test_convert_dtypes.py]: Sounds good. Like this?

File: pandas/tests/series/methods/test_convert_dtypes.py [hunk 2]
```diff
@@ -225,3 +227,16 @@ def test_convert_bool_dtype(self):
         # GH32287
         df = pd.DataFrame({"A": pd.array([True])})
         tm.assert_frame_equal(df, df.convert_dtypes())
+
+    def test_convert_object_to_datetime_with_cache(self):
```
Comment (jreback) [inline pandas/tests/series/methods/test_convert_dtypes.py:231]: can you put this with the other to_datetime tests: pandas/tests/tools/test_to_datetime.py; there a number of caching tests if you can locate nearby

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/series/methods/test_convert_dtypes.py (hunk 2 above)
Comment (phofl) [reply pandas/tests/series/methods/test_convert_dtypes.py]: Moved, hope the location is ok, cache tests are everywhere there :)

**Comment by phofl** [comment]
@jreback greenish

**Review by jreback — APPROVED** [review]

File: pandas/tests/series/methods/test_convert_dtypes.py (hunk 2 above)
Comment (jreback) [reply pandas/tests/series/methods/test_convert_dtypes.py]: yep lgtm. lots of tests :->

**Comment by jreback** [comment]
thanks @phofl

**Pull request merged**